### ISSUE TITLE
Only use one network request when identical requests are made

### ISSUE THREAD

**Issue by jeswr** [nodejs/undici#3232]
## This would solve...

Performance and robustness issues associated with fetching the same HTTP resource in parallel in several places across a codebase.

## The implementation should look like...

Using an interceptor, keep an internal record of pending requests, if an identical request is already pending re-use the response across the multiple pending requests

## Additional context

Originally raised in https://github.com/nodejs/undici/issues/3221

**Comment by mcollina** [comment]
The hardest part of this is what is considered an _indentical_ request.

Overall, this would be a nice interceptor to add.

**Comment by metcoder95** [comment]
I'd suggest starting with an idempotent request by default and its request for fingerprint for checking _identity_, and verifying its response with a possible timeout to invalidate the _cached responses_.

Also some hooks similar to what we did for `retry` interceptor so implementations can customize the behaviour for validating if two requests are identical.

**Comment by Uzlopak** [comment]
In rfc 9111 it actually defines requests where the http cache should return the same cached response. This definition could be used.

**Comment by Uzlopak** [comment]
Btw. How do you handle streams?

**Comment by metcoder95** [comment]
True, this can help provide some guidance for defaults: https://www.rfc-editor.org/rfc/rfc9111.html#name-storing-responses-in-caches

I'd just still be mindful of providing some control to the implementation.

>Btw. How do you handle streams?

wdym?

**Comment by ronag** [comment]
With streams you can have multiple consumers consuming the response with different speeds. We should only do this for requests where the response body can be fully buffered into memory.

**Comment by blephy** [comment]
I wrote a code and push it in production addressing this issue but only for auhentification : if one app request simultaneously 2 micro-services connected to the same auth backend, only one authentification is made.

It's not an hard part to do it, but there is a lot of things to consider when whe are talking about "same request"... 

Additionally, when a GET request is made, this is often the authentication which are taking a lot of time who are interested to be shared. Without this in mind, the cache interceptor can be used after an auth interceptor, but with caution.

**Comment by mcollina** [comment]
Please try https://github.com/nodejs/undici/pull/4679 and report back
